flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesClusterDescriptor.java

```diff
@@ -0,0 +1,190 @@
+
+	public KubernetesClusterDescriptor(Configuration flinkConfig, FlinkKubeClient client) {
+		this.flinkConfig = flinkConfig;
+		this.client = client;
+		this.clusterId = flinkConfig.getString(KubernetesConfigOptions.CLUSTER_ID);
+		Preconditions.checkNotNull(clusterId, "ClusterId must be specified!");
+	}
+
+	@Override
+	public String getClusterDescription() {
+		return CLUSTER_DESCRIPTION;
+	}
+
+	private ClusterClient<String> createClusterClient(String clusterId) throws Exception {
+
+		Configuration configuration = new Configuration(flinkConfig);
+
+		final Endpoint restEndpoint = this.client.getRestEndpoint(clusterId);
+
+		if (restEndpoint != null) {
+			configuration.setString(RestOptions.ADDRESS, restEndpoint.getAddress());
+			configuration.setInteger(RestOptions.PORT, restEndpoint.getPort());
+		} else {
+			throw new ClusterRetrieveException("Could not get the rest endpoint of " + clusterId);
+		}
+
+		return new RestClusterClient<>(configuration, clusterId);
+	}
+
+	@Override
+	public ClusterClient<String> retrieve(String clusterId) throws ClusterRetrieveException {
+		try {
+			final ClusterClient<String> retrievedClient = createClusterClient(clusterId);
+			LOG.info("Retrieve flink cluster {} successfully, JobManager Web Interface : {}", clusterId,
+				retrievedClient.getWebInterfaceURL());
+			return retrievedClient;
+		} catch (Exception e) {
+			this.client.handleException(e);
+			throw new ClusterRetrieveException("Could not create the RestClusterClient.", e);
+		}
+	}
+
+	@Override
+	public ClusterClient<String> deploySessionCluster(ClusterSpecification clusterSpecification)
+		throws ClusterDeploymentException {
+
+		final ClusterClient<String> clusterClient = deployClusterInternal(
+			KubernetesSessionClusterEntrypoint.class.getName(), clusterSpecification, false);
+
+		LOG.info("Create flink session cluster {} successfully, JobManager Web Interface: {}",
+			clusterId, clusterClient.getWebInterfaceURL());
+		return clusterClient;
+	}
+
+	@Override
+	public ClusterClient<String> deployJobCluster(ClusterSpecification clusterSpecification,
+			JobGraph jobGraph, boolean detached) throws ClusterDeploymentException {
+		throw new ClusterDeploymentException("Per job could not be supported now.");
+	}
+
+	private ClusterClient<String> deployClusterInternal(
+			String entryPoint,
+			ClusterSpecification clusterSpecification,
+			boolean detached) throws ClusterDeploymentException {
+
+		final ClusterEntrypoint.ExecutionMode executionMode = detached ?
+			ClusterEntrypoint.ExecutionMode.DETACHED
+			: ClusterEntrypoint.ExecutionMode.NORMAL;
+		flinkConfig.setString(ClusterEntrypoint.EXECUTION_MODE, executionMode.toString());
+
+		flinkConfig.setString(KubernetesConfigOptionsInternal.ENTRY_POINT_CLASS, entryPoint);
+
+		// Rpc(6123), blob(6124), rest(8081) taskManagerRpc(6122) port need to be exposed, so update them to fixed port.
+		if (Integer.valueOf(flinkConfig.get(BlobServerOptions.PORT)) == 0) {
```

Comment:
Possibly we catch the parse exception and throw/log a specific information that port range doesn't support in K8S deployment. And when we config the default port on user configure `0`, it'd better we log a warning so that user knows his original purpose "random choose  a port" isn't respected.

```java
        {
			final int blobServerPort;
			try {
				blobServerPort = Integer.parseInt(flinkConfig.get(BlobServerOptions.PORT));
			} catch (NumberFormatException e) {
				// log...
				throw new ClusterDeploymentException("...");
			}
			
			if (blobServerPort == 0) {
				flinkConfig.setString(BlobServerOptions.PORT, String.valueOf(Constants.BLOB_SERVER_PORT));
				// log ...
			}
		}

		{
			final int taskManagerRpcPort;
			try {
				taskManagerRpcPort = Integer.parseInt(flinkConfig.get(BlobServerOptions.PORT));
			} catch (NumberFormatException e) {
				// log...
				throw new ClusterDeploymentException("...");
			}

			if (taskManagerRpcPort == 0) {
				flinkConfig.setString(TaskManagerOptions.RPC_PORT, String.valueOf(Constants.TASK_MANAGER_RPC_PORT));
				// log ...
			}
		}
```

Reply:
I will add a method `parsePort` in `KubernetesUtils` so that it could be reused by `FlinkMasterDeploymentDecorator` and `TaskManagerPodDecorator`.